storage/redis_adapter.go

```diff
@@ -3,13 +3,67 @@ package storage
 import (
 	"fmt"
 	"log"
+	"time"
 
 	r "gopkg.in/redis.v3"
 )
 
+type message struct {
+	app         string
+	messageBody string
+}
+
+func newMessage(app string, messageBody string) *message {
+	return &message{
+		app:         app,
+		messageBody: messageBody,
+	}
+}
+
+type messagePipeliner struct {
+	bufferSize    int
+	messageCount  int
+	pipeline      *r.Pipeline
+	timeoutTicker *time.Ticker
+	queuedApps    map[string]bool
+}
+
+func newMessagePipeliner(bufferSize int, redisClient *r.Client) *messagePipeliner {
+	return &messagePipeliner{
+		bufferSize:    bufferSize,
+		pipeline:      redisClient.Pipeline(),
+		timeoutTicker: time.NewTicker(time.Second),
+		queuedApps:    map[string]bool{},
+	}
+}
+
+func (mp *messagePipeliner) addMessage(message *message) {
+	if err := mp.pipeline.RPush(message.app, message.messageBody).Err(); err == nil {
+		mp.queuedApps[message.app] = true
+		mp.messageCount++
+	} else {
+		log.Printf("Error adding rpush to %s to the pipeline: %s", message.app, err)
+	}
+}
+
+func (mp *messagePipeliner) execPipeline() {
```

Comment:
in most cases, a non-pointer receiver will deep-copy, yes. in this case, the deep-copy will be cheap (not as cheap as a pointer, obviously) because the struct looks like it contains references or pointer-sized (approximately) values.

also, there is no race afaik - I was referring to being defensive against races in the future.